### ISSUE TITLE
ttnn.typecast uint32 -> float32 rounds twice for inputs >= 2^31: 353,697,792 of them (16.47%) come back 1 ULP away from torch's correctly rounded value

### ISSUE THREAD

**Issue by Yusaku-062255** [tenstorrent/tt-metal#57643]
**Component / Area:** kernels (LLK/SFPU `ckernel_sfpu_typecast.h`) and `ttnn.typecast`

**Issue Type:** Bad Outputs

## Summary

`ttnn.typecast(x_uint32, ttnn.float32)` does not round correctly for inputs with bit 31 set. The kernel converts the low 31 bits to fp32 (one round-to-nearest-even), then adds 2^31 (a second round-to-nearest-even). Two roundings are not the same as one. For 353,697,792 of the 2,147,483,648 inputs in [2^31, 2^32), which is 16.47% of the upper half and 8.24% of the whole uint32 range, the result is the fp32 neighbour of the correctly rounded value. `torch.Tensor.to(torch.float32)` on a `torch.uint32` tensor rounds once and gets the nearest value.

```text
v = 3221225663 (= 2^31 + 2^30 + 191), uint32 -> float32

ttnn.typecast (replay)      3221225472.0   |v - out| = 191   (more than half an ULP: not the nearest float)
torch .to(torch.float32)    3221225728.0   |v - out| = 65
```

The error is always exactly 1 ULP (256 at this magnitude). It is silent, it does not depend on the input distribution beyond "bit 31 set", and it breaks the op's own contract: the unit test for this exact dtype pair compares with `assert_equal`, i.e. it requires bit-exact agreement with `x.to(torch.float32)`.

This is a source-level analysis: I do not have a card. The replay below follows the kernel instruction by instruction, runs exhaustively over all 2^31 affected inputs, and needs no device.

## Root cause

[`ckernel_sfpu_typecast.h#L443-L449`](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_typecast.h#L443-L449) (Blackhole, `DISABLE_SFPLOADMACRO` path; Wormhole is the same at [L459-L465](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_typecast.h#L459-L465)):

```cpp
        TTI_SFPLOAD(p_sfpu::LREG0, InstrModLoadStore::INT32, ADDR_MOD_7, 0);
        TTI_SFPSETSGN(0, p_sfpu::LREG0, p_sfpu::LREG1, 1);                  // lo = v with bit 31 cleared
        TTI_SFPCAST(p_sfpu::LREG1, p_sfpu::LREG2, 0);                       // f = fp32(lo), rounding #1 (RNE)
        TTI_SFPSETCC(0, p_sfpu::LREG0, 0, sfpi::SFPSETCC_MOD1_LREG_LT0);    // lanes with bit 31 set
        TTI_SFPADDI(0x4f00, p_sfpu::LREG2, 0);  // 2^31                      // f = f + 2^31, rounding #2 (RNE)
        TTI_SFPENCC(0, 0, 0, 0);
        TTI_SFPSTORE(p_sfpu::LREG2, InstrModLoadStore::FP32, ADDR_MOD_6, 0);
```

The default SFPLOADMACRO path does the same thing with the same two roundings: `SFPCAST(a, 13, 0)` then `SFPMAD(L[L7], 1.0, v)` with `L1 = 2^31` ([BH init L659-L665](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_typecast.h#L659-L665), [macro schedule L451-L489](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_typecast.h#L451-L489); [WH init L681-L687](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_typecast.h#L681-L687)). `SFPCAST` mode 0 is `SM32_TO_FP32_RNE` (named so in [`ckernel_sfpu_exp.h#L243-L244`](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_exp.h#L243-L244)).

Why two roundings go wrong: for `lo >= 2^24` the cast rounds `lo` to a multiple of 2^(e-23) (up to 128 for `lo >= 2^30`). Adding 2^31 moves the value into [2^31, 2^32), where the fp32 spacing is 256, so it rounds again. When the first rounding lands exactly on an odd multiple of 128, the second rounding sees a tie that did not exist in the exact value and breaks it toward even, which can be the far side. Worked example, `v = 2^31 + 2^30 + 191`:

- exact: `191` above `2^31 + 2^30` rounds to `+256` (distance 65, versus 191) -> `3221225728`
- device: `cast(2^30 + 191)` rounds 191 to 128 (spacing 128), then `2^31 + 2^30 + 128` is a tie between `+0` and `+256`, RNE picks the even mantissa `+0` -> `3221225472`

`ttnn.typecast` reaches this kernel through `typecast_tile<UInt32, Float32>` ([`typecast.h#L256-L263`](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tt_metal/hw/inc/api/compute/eltwise_unary/typecast.h#L256-L263)). Inputs below 2^31 skip the add and round once, so they are exact-correct.

## Reference semantics

`torch.tensor(v, dtype=torch.uint32).to(torch.float32)` is a single C++ `static_cast<float>` of the integer, correctly rounded to nearest-even (checked with torch 2.8: `3221225663 -> 3221225728.0`). numpy `astype(np.float32)` agrees. The tt-metal golden for this pair is `x.to(torch.float32)` ([`ttnn_pytorch_ops.py#L268-L269`](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tests/ttnn/python_api_testing/sweep_tests/ttnn_pytorch_ops.py#L268-L269)), compared bit-exactly.

## Impact

Exhaustive replay over every uint32 with bit 31 set (script below):

| low 31 bits `lo` | inputs | wrong (1 ULP) | share |
|---|---:|---:|---:|
| < 2^24 | 16,777,216 | 0 | 0% |
| [2^24, 2^25) | 16,777,216 | 65,536 | 0.39% |
| [2^25, 2^26) | 33,554,432 | 262,144 | 0.78% |
| [2^26, 2^27) | 67,108,864 | 1,048,576 | 1.56% |
| [2^27, 2^28) | 134,217,728 | 4,194,304 | 3.13% |
| [2^28, 2^29) | 268,435,456 | 16,777,216 | 6.25% |
| [2^29, 2^30) | 536,870,912 | 67,108,864 | 12.50% |
| [2^30, 2^31) | 1,073,741,824 | 264,241,152 | 24.61% |
| **all of [2^31, 2^32)** | **2,147,483,648** | **353,697,792** | **16.47%** |

- The wrong results split evenly up and down (176,848,896 each), so the error does not cancel in a sum.
- Every wrong result is more than half an ULP from the exact integer, so it is not "a different but valid rounding". It is simply not the nearest float.
- uint32 values at or above 2^31 are what a hash, a random-bit generator, a bit-packed mask or a large index/counter produce. Converting those to float32 for scaling or normalisation silently disagrees with torch and numpy.
- Wormhole and Blackhole, both SFPLOADMACRO and `DISABLE_SFPLOADMACRO` builds.
- Side effect on uint32 -> bfloat16 ([BH L384-L436](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_typecast.h#L384-L436)), which uses the same cast + add before the bf16 rounding: an exhaustive replay (modelling the final fp32 -> bf16 step as RNE) finds 5,397 inputs in [2^31, 2^32) where it differs from torch's `float(v)` -> bf16. The fix below makes those agree too.

### Why it survived

The test for this pair demands exact equality but never sets bit 31. `test_run_eltwise_typecast_op` includes `(torch.int, ttnn.uint32, ttnn.float32)` ([`test_eltwise_typecast.py#L82`](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tests/ttnn/unit_tests/operations/eltwise/test_eltwise_typecast.py#L82)) and checks float outputs with `assert_equal` ([L183-L184](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tests/ttnn/unit_tests/operations/eltwise/test_eltwise_typecast.py#L183-L184)), but `typecast_test_input_bounds` returns `(0, 100)` for any non-integer output dtype ([`typecast_test_helpers.py#L66-L67`](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tests/ttnn/python_api_testing/typecast_test_helpers.py#L66-L67)), and uint32 test inputs are built as `torch.int32` and capped at `INT32_MAX` ([L22-L23](https://github.com/tenstorrent/tt-metal/blob/9c57134a0d7f328e158daf71390ede0f3694fde9/tests/ttnn/python_api_testing/typecast_test_helpers.py#L22-L23)). Values below 2^24 convert exactly under any rounding, so the test cannot fail.

## Proposed fix

Round once. For lanes with bit 31 set, halve with a sticky bit, cast, and double:

```text
h = (v >> 1) | (v & 1)     // < 2^31, so SFPCAST sees a positive sign-magnitude integer
f = SFPCAST(h)             // the only rounding (RNE); the rounding point is bit 7 or higher of h,
                           // so the one sticky bit carries everything that was shifted out
f = addexp(f, +1)          // x2, exact (SFPDIVP2 / exponent + 1)
```

Lanes below 2^31 keep the current single `SFPCAST`. The replay checks this over all 2^31 affected inputs: 0 mismatches against torch. In the `DISABLE_SFPLOADMACRO` loop this replaces `SFPSETSGN` + `SFPADDI` under the existing `SFPSETCC` with a shift, an OR and a `SFPDIVP2`. The macro path needs the same change in its templates, or can fall back to the loop for this one conversion if the extra slot does not fit the 3-cycle schedule. The same change fixes uint32 -> bfloat16, which shares the cast + add.

Add a regression case that actually sets bit 31: a `torch.uint32` input with values such as `3221225663, 4107611324, 2971424921, 4294967295` plus a random sample over [2^31, 2^32), compared with `assert_equal` against `x.to(torch.float32)`.

Happy to prepare the PR.

## Reproducer

Host only, no device required (`pip install numpy torch`, torch >= 2.3 for `torch.uint32`). Runs in about 20 s:

```python
"""Host-side replay of calculate_typecast_uint32_to_fp32() for uint32 inputs >= 2^31.

Replays tt_metal/hw/ckernels/{wormhole_b0,blackhole}/metal/llk_api/llk_sfpu/ckernel_sfpu_typecast.h
(tt-metal main @ 9c57134a0d7f; same arithmetic on both archs and on both the SFPLOADMACRO and the
DISABLE_SFPLOADMACRO paths). Host only: no Tenstorrent device needed. Requires numpy and torch.

Kernel, per element v (uint32):
    lo = setsgn(v, 0)          # clear bit 31                          (SFPSETSGN)
    f  = cast(lo)              # sign-magnitude int -> fp32, RNE       (SFPCAST mod 0 = SM32_TO_FP32_RNE)
    if bit31(v): f = f + 2^31  # fp32 add, RNE                          (SFPADDI 0x4f00 / SFPMAD indirect VA)
Reference: torch.tensor(v, dtype=torch.uint32).to(torch.float32), i.e. one RNE rounding of the exact integer.
"""
import numpy as np
import torch

TWO31 = np.float32(2.0**31)


def device(lo_int64):
    """lo = v - 2^31 (v has bit 31 set). Two roundings: cast, then add."""
    return lo_int64.astype(np.float32) + TWO31


def reference(v_int64):
    return v_int64.astype(np.float32)  # single RNE rounding (exact int -> fp32)


# 1. worked examples, checked against torch's own uint32 -> float32 conversion
examples = [3221225663, 4107611324, 3714065583, 2971424921, 2147483648 + 193, 4294967295]
t = torch.tensor(examples, dtype=torch.uint32).to(torch.float32).to(torch.float64).tolist()
print("     v (uint32)   exact - 2^31 | device (2 roundings)   torch.to(float32) | |v-dev| |v-torch|")
for v, tv in zip(examples, t):
    d = float(device(np.array([v - 2**31], dtype=np.int64))[0])
    r = float(reference(np.array([v], dtype=np.int64))[0])
    assert r == tv, (v, r, tv)
    flag = "  <-- 1 ULP off" if d != r else ""
    print(f"{v:>15d} {v - 2**31:>14d} | {d:>20.1f} {tv:>19.1f} | {abs(v - d):>7.0f} {abs(v - tv):>8.0f}{flag}")

# 2. exhaustive count over all 2^31 inputs with bit 31 set
CH = 1 << 26
total = up = down = worst = 0
bins = {}
for start in range(0, 1 << 31, CH):
    lo = np.arange(start, start + CH, dtype=np.int64)
    dev, ref = device(lo), reference(lo + (1 << 31))
    m = dev != ref
    n = int(m.sum())
    if n:
        diff = dev[m].astype(np.float64) - ref[m].astype(np.float64)
        up += int((diff > 0).sum()); down += int((diff < 0).sum())
        worst = max(worst, int(np.abs(diff).max()))
        k = int(np.log2(start)) if start else 0
        bins[k] = bins.get(k, 0) + n
    total += n
print(f"\nall uint32 in [2^31, 2^32): {total:,} / {1 << 31:,} = {total / (1 << 31):.4%} differ from torch "
      f"(rounded up {up:,}, down {down:,}); max |error| = {worst} = {worst // 256} ULP (ULP = 256 here)")
print(f"as a share of the whole uint32 range: {total / (1 << 32):.4%}")

# 3. control: inputs below 2^31 take one rounding and always match
lo = np.arange(0, 1 << 31, 997, dtype=np.int64)
print("control, v < 2^31 (every 997th value):", int((lo.astype(np.float32) != reference(lo)).sum()), "mismatches")

# 4. the proposed single-rounding fix: cast((v >> 1) | (v & 1)) then scale by 2 (exponent + 1)
def fixed(v):
    h = (v >> 1) | (v & 1)       # 31-bit value with a sticky bit: one RNE rounding in the cast is exact-equivalent
    return h.astype(np.float32) * np.float32(2.0)
bad = 0
for start in range(0, 1 << 31, CH):
    v = np.arange(start, start + CH, dtype=np.int64) + (1 << 31)
    bad += int((fixed(v) != reference(v)).sum())
print("proposed fix (sticky-halve, cast, addexp +1): mismatches over [2^31, 2^32):", bad)
```

Output:

```text
     v (uint32)   exact - 2^31 | device (2 roundings)   torch.to(float32) | |v-dev| |v-torch|
     3221225663     1073742015 |         3221225472.0        3221225728.0 |     191       65  <-- 1 ULP off
     4107611324     1960127676 |         4107611136.0        4107611392.0 |     188       68  <-- 1 ULP off
     3714065583     1566581935 |         3714065408.0        3714065664.0 |     175       81  <-- 1 ULP off
     2971424921      823941273 |         2971424768.0        2971425024.0 |     153      103  <-- 1 ULP off
     2147483841            193 |         2147483904.0        2147483904.0 |      63       63
     4294967295     2147483647 |         4294967296.0        4294967296.0 |       1        1

all uint32 in [2^31, 2^32): 353,697,792 / 2,147,483,648 = 16.4703% differ from torch (rounded up 176,848,896, down 176,848,896); max |error| = 256 = 1 ULP (ULP = 256 here)
as a share of the whole uint32 range: 8.2352%
control, v < 2^31 (every 997th value): 0 mismatches
proposed fix (sticky-halve, cast, addexp +1): mismatches over [2^31, 2^32): 0
```

(The replay treats `SFPCAST` mode 0 and the `SFPADDI` / `SFPMAD` add as IEEE fp32 round-to-nearest-even, which is what the mode name and the kernel's own "2**31" comments imply. The last row of the first table, `4294967295 -> 4294967296.0`, is the correct round-up to 2^32 and matches torch.)

A device check, for anyone with a card (I could not run it):

```python
import torch, ttnn
device = ttnn.open_device(device_id=0)
v = torch.tensor([3221225663, 4107611324, 3714065583, 2971424921] * 256, dtype=torch.uint32).reshape(1, 1, 32, 32)
tt = ttnn.from_torch(v, dtype=ttnn.uint32, layout=ttnn.TILE_LAYOUT, device=device)
out = ttnn.to_torch(ttnn.typecast(tt, ttnn.float32))
print(out[0, 0, 0, :4].double())             # replay predicts [3221225472, 4107611136, 3714065408, 2971424768]
print(v[0, 0, 0, :4].to(torch.float32).double())  # torch:          [3221225728, 4107611392, 3714065664, 2971425024]
```

**Comment by Yusaku-062255** [comment]
For the record, as the reporter: I am submitting this as a community-suggested bounty to bounties@tenstorrent.com, as the maintainers asked on #55931, and would like to be assigned before any PR is written, per the Bounty Program Terms. The single-rounding fix (sticky-halve, cast, exponent +1) in the "Proposed fix" section is verified with 0 mismatches over all 2^31 affected inputs in the replay, and the regression cases are ready on my side. @jberkowitzTT could you review this as a bounty candidate?
